src/sinks/util/service/health.rs

```diff
@@ -0,0 +1,348 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+use futures::{ready, FutureExt};
+use futures_util::{future::BoxFuture, TryFuture};
+use pin_project::pin_project;
+use tokio::time::{sleep, Duration};
+use tower::Service;
+use vector_config::configurable_component;
+
+use crate::{
+    emit,
+    internal_events::{EndpointsActive, OpenGauge, OpenToken},
+    sinks::util::retries::ExponentialBackoff,
+};
+
+const RETRY_MAX_DURATION_SECONDS_DEFAULT: u64 = 3_600;
+const RETRY_INITIAL_BACKOFF_SECONDS_DEFAULT: u64 = 1;
+const UNHEALTHY_AMOUNT_OF_ERRORS: usize = 5;
+/// How many errors is allowed in probation period before we consider the service unhealthy again.
+/// Must be less than UNHEALTHY_AMOUNT_OF_ERRORS and greater than 0.
+const PROBATION_AMOUNT: usize = 4;
+
+/// Options for determining health of an endpoint.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+#[serde(rename_all = "snake_case")]
+pub struct HealthConfig {
+    /// Initial timeout, in seconds, between attempts to reactivate endpoints once they become unhealthy.
+    pub retry_initial_backoff_secs: Option<u64>,
+
+    /// Maximum timeout, in seconds, between attempts to reactivate endpoints once they become unhealthy.
+    pub retry_max_duration_secs: Option<u64>,
+}
+
+impl HealthConfig {
+    pub fn build<S, L>(
+        &self,
+        logic: L,
+        inner: S,
+        healthcheck: impl Fn() -> BoxFuture<'static, bool> + Send + 'static,
+        open: OpenGauge,
+        endpoint: String,
+    ) -> HealthService<S, L> {
+        let counters = Arc::new(HealthCounters::new());
+        let snapshot = counters.snapshot();
+
+        HealthService {
+            inner,
+            logic,
+            counters,
+            snapshot,
+            open,
+            endpoint,
+            state: ServiceState::Healthcheck(healthcheck()),
```

Comment:
I think with current healthcheck logic we'd still send events to a failed sink, but depending on the options passed to Vector it may exit on a failed check.

Reply:
Hmmm, what options would those be?

I thought that enabling healthchecks at all would run the healthcheck at start before the topology was marked as ready, and if it passed, things would proceed, and if not, Vector would crash, as you've described.

Like, the sink could go unhealthy between checking and Vector saying it's ready, sure... but you'd have still checked prior to that and gotten a "we're healthy" response before trying that first event in the sink, no?

Reply:
[VECTOR_REQUIRE_HEALTHY](https://vector.dev/docs/reference/cli/#vector_require_healthy) is optional and causes Vector to exit on any healthcheck fail. So the default behavior is "check health, log error if not, send events", and the check is never run again.

Reply:
Hmmm, fair enough.

Reply:
Yeah, there's some outstanding issues to improve the healthchecks because they are very naive. I don't have a strong feeling for how this particular PR changes things, but I wanted to make sure we had a shared understanding of the behavior today. 👍

Reply:
As Spencer mentioned, by default sink can run with unhealthy endpoints but even if that wasn't the case I think health service is more robust if it isn't assuming health of endpoints. An example of a case breaking that health assumption is snooping.